### DISCUSSION TITLE
Strawberryshake and singletons in dependency injection

### DISCUSSION THREAD

**Discussion by antyrama** [ChilliCream/hotchocolate#5452] [Q&A] [answered]
Hi,

I'm building multitenant solution with `StrawberryShake.CodeGeneration.CSharp.Analyzers 12.14.0` and have an issue as everything there is being registered into IoC as singleton so there is no way (or maybe I'm mistaken) to have multiple instances with different `HttpClient.BaseAddress` per tenant.

Is there a way to generate code with scoped lifestyle of the objects?

Cheers
P

**Comment by michaelstaib** [comment, marked as answer]
This will be coming in a couple of builds ....

**Reply by antyrama** [reply]
Thanks. For now we have workaround, for anybody that may have similar issue:
```
builder.Services
    .AddSingleton<IHttpContextAccessor, HttpContextAccessor>()
    .AddScoped<TenantContext>()
    .AddScoped<ICrystallizeClient>(provider =>
    {
        var tenantContext = provider.GetRequiredService<TenantContext>();

        var childBuilder = WebApplication.CreateBuilder();

        childBuilder.Services.AddCrystallizeClient()
            .ConfigureHttpClient(client =>
            {
                client.BaseAddress = new Uri($"https://api.crystallize.com/{tenantContext.Name}/catalogue");
            });

        return childBuilder.Build()
            .Services
            .GetRequiredService<ICrystallizeClient>();
    });
```

**Reply by tobias-tengler** [reply]
I don't think you need to build a full WebApplication everytime. Just do

```csharp
new ServiceCollection()
  .AddClient...()
  .BuildServiceProvider()
  .GetRequiredService<...>();
```

(Sorry for the incomplete code, am on mobile)

**Reply by antyrama** [reply]
Yep, you're right 👍, the above snipped was just first shot
